diff -rupN mawk-1.3.4-20130803.orig/bi_vars.c mawk-1.3.4-20130803/bi_vars.c
--- mawk-1.3.4-20130803.orig/bi_vars.c	2013-08-31 13:22:51.000000000 -0700
+++ mawk-1.3.4-20130803/bi_vars.c	2013-08-31 13:23:08.000000000 -0700
@@ -78,7 +78,7 @@ bi_vars_init(void)
     FILENAME->ptr = (PTR) new_STRING("");
 
     OFS->type = C_STRING;
-    OFS->ptr = (PTR) new_STRING(" ");
+    OFS->ptr = (PTR) new_STRING("\t");
 
     ORS->type = C_STRING;
     ORS->ptr = (PTR) new_STRING("\n");
diff -rupN mawk-1.3.4-20130803.orig/field.c mawk-1.3.4-20130803/field.c
--- mawk-1.3.4-20130803.orig/field.c	2013-08-31 13:22:51.000000000 -0700
+++ mawk-1.3.4-20130803/field.c	2013-08-31 13:40:41.000000000 -0700
@@ -189,8 +189,8 @@ field_init(void)
 
     load_pfield("FS", FS);
     FS->type = C_STRING;
-    FS->ptr = (PTR) new_STRING(" ");
-    /* fs_shadow is already set */
+    FS->ptr = (PTR) new_STRING("\t");
+    cast_for_split(cellcpy(&fs_shadow, FS));
 
     load_pfield("OFMT", OFMT);
     OFMT->type = C_STRING;
